### DISCUSSION TITLE
General notes

### DISCUSSION THREAD

**Discussion by ShataDg** [broadinstitute/MicroSplit-reproducibility#5] [General]
#### Additional notes related to `microsplit` [experiments](https://github.com/broadinstitute/MicroSplit-reproducibility/discussions/6)

Changing (reducing) the batch size, changes (increases) the number of steps in an epoch

<img width="600" height="400" alt="image" src="https://github.com/user-attachments/assets/8356aeb5-bf7f-4420-8994-d168595f7a59" />

<img width="1924" height="116" alt="image" src="https://github.com/user-attachments/assets/9fc33684-933a-42e3-9454-3dec46007a96" />

Changing the training set (reducing number of images) reduces the number of steps in an epoch (batch and patch sizes and number of epochs were kept the same)

<img width="989" height="286" alt="image" src="https://github.com/user-attachments/assets/37dc37c9-ebb6-4579-aebf-474aeb6ba2de" />

<img width="1356" height="59" alt="image" src="https://github.com/user-attachments/assets/998aeec1-6af7-4dd9-8fb5-208820bce0de" />

**Comment by ShataDg** [comment]
## Prediction and calibration

`02_predict.ipynb` and `03_calibration.ipynb`

1. Runs the trained model on new images
2. Output: predicted unmixed channels
3. Outputs are sampled from learned posterior predictions --> outputs are probabilistic
4. Computes uncertainty heatmaps (check)
5. Compare predictions to ground truths (SSIM, PSNR, MSE, MAE, cosine similarity)
6. Image is evaluated at three image scales `multiscale_count=3`
7. Predictions are made on validation and held-out datasets
8. Assignment into train and val data is being done by `val_data_config` and `train_data_config` --> perhaps modify the rules so that a curated set of validation and held-out datasets can be passed here
9. Inference runs on image tiles and stitches them back together
10. Inner padding is performed to avoid seam artifacts at tile borders. For each tile, the prediction is done on a slightly bigger region. Then the unreliable border pixels are discarded and the “inner” region is kept for stitching. As the tiles overlap, every pixel in the final image comes from a region where the model had enough context.
11. `mmse_count` sets the number of posterior samples per tile; these are averaged to produce the MMSE estimate. Higher values of `mmse_count` reduces random variation and ensures smoother stitching.
12. The following step shows two posterior samples and their differences
```
from microsplit_reproducibility.notebook_utils.HT_LIF24 import show_sampling

imgsz = 3
ncols = 6
examplecount = 3
_, ax = plt.subplots(
    figsize=(imgsz * ncols, 2 * imgsz * examplecount),
    ncols=ncols,
    nrows=2 * examplecount,
)

show_sampling(dset, model, ax=ax[:2])
show_sampling(dset, model, ax=ax[2:4])
show_sampling(dset, model, ax=ax[4:6])
plt.tight_layout()
```
13. Files produced in the `02_predict.ipynb` notebook:
    1. **Target** – ground-truth single-channel images  
    2. **Std** – pixel-wise standard deviation from the posterior samples  
    3. **Pred** – MMSE predictions (stitched mean prediction)

14. `03_calibration.ipynb` measures the actual error (`target` - `pred`) and compares it against predicted uncertainty 
15. Metrics computed
- **Per-pixel error:** `error = pred - target` or `abs(pred - target)`
- **RMSE per bin:** root-mean-square error of pixels in the bin
- **RMV per bin (or similar):** root-mean variance / predicted uncertainty summary derived from `pred_std` (`sqrt(mean(std^2))` or `mean(std)`)
